### PR TITLE
fix: pass O_SHARE_DELETE when digesting files

### LINKED ISSUE #8268: background digests crash on Windows
> It's a solution in the sense that it prevents dune from crashing. I'll open a new issue to track "let's reenable background digests on windows". (and I'll close this one otherwise it'll keep being closed whenever a branch containing the fix is pushed, sorry about that)

_Originally posted by @emillon in https://github.com/ocaml/dune/issues/8228#issuecomment-1649486223_

### PR DESCRIPTION
**Pull request by emillon**
Fixes #8268

This flag corresponds to what `Stdlib.open_in` does. When passed, the files can be removed while they are still open. This condition can happen when background digests are enabled. This is a no-op everywhere else.

This allows enabling background digests on Windows.

### CODE REVIEW

**Review by rgrinberg — COMMENTED** [review]
Can we leave the old code on Unix? Channels have mutexes that we don't really need here

**Comment by emillon** [comment]
I'll try the other following approaches:
- use Unix.open with `_O_NOINHERIT`
- replace close by `CloseHandle(Handle_val(v))`

**Comment by emillon** [comment]
Ah, even better: adding `O_SHARE_DELETE` at the open point fixes the issue (and is a noop on nonwindows). I'll update the PR.

**Comment by emillon** [comment]
There's still a part I don't completely understand (why the file is closed too late in some cases) but it might be related to the details of threading and win32unix.

File: src/dune_digest/dune_digest.ml [hunk 1]
```diff
@@ -18,7 +18,7 @@ end
 module Direct_impl : Digest_impl = struct
   let file file =
     let fd =
-      match Unix.openfile file [ Unix.O_RDONLY ] 0 with
+      match Unix.openfile file [ Unix.O_RDONLY; O_SHARE_DELETE ] 0 with
```
Comment (hhugo) [inline src/dune_digest/dune_digest.ml:21]: Shoudn't this come with a comment ? maybe referencing the gh issue ..

**Review by nojb — APPROVED** [review]
LGTM. Understanding why the file is not closed in the first place would be ideal of course, but this fix looks solid.

**Comment by emillon** [comment]
I added a comment explaining the situation. let me know what you think.

**Comment by nojb** [comment]
> I added a comment explaining the situation. let me know what you think.

LGTM

**Comment by hhugo** [comment]
LGTM

**Pull request merged**